### PR TITLE
Whimsical Woodpeckers: Whimsy Messaging

### PR DESCRIPTION
**Pull request by pmdevita**

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
One small thing that needs to be handled during merge: `.gitignore` file change.

No flake8 lint errors

**Comment by pmdevita** [comment]
I undid our .gitignore changes, sorry about that

**Comment by SebastiaanZ** [comment]
> 
> 
> I undid our .gitignore changes, sorry about that

No worries, that was not an issue at all. Thanks for the quick fix, though.

File: whimsical-woodpeckers/README.md [hunk 1]
```diff
@@ -1,2 +1,40 @@
 # Whimsical Woodpeckers
```
Comment (eivl) [inline whimsical-woodpeckers/README.md:1]: Thanks for participating!

File: whimsical-woodpeckers/README.md [hunk 2]
```diff
@@ -1,2 +1,40 @@
...
 This is the default README of your team's project. Please replace this by a README with more information on your project. At the very least, your README should contain information on how to set-up and run your project.
+
```
Comment (eivl) [inline whimsical-woodpeckers/README.md:3]: Clear README, easy to follow. looks like there are missing the command to migrate the database.

File: whimsical-woodpeckers/README.md [hunk 3]
```diff
@@ -1,2 +1,40 @@
...
+```
+### Include a .env file or set environment variables to the following
+```
+SECRET_KEY=YOUR-SECRET-KEY
+DEBUG=False
+```
+
+### Run Django
+```shell
+pipenv run python manage.py runserver
```
Comment (eivl) [inline whimsical-woodpeckers/README.md:39]: pipenv run python manage.py migrate

File: whimsical-woodpeckers/anon/models.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    auth_token = models.CharField(max_length=30, default="")
+    auth_expiration = models.DateTimeField(default=timezone.now)
+
+    USERNAME_FIELD = 'id'
+
+    def __str__(self):
+        return self.nickname
+
+    def save(self, *args, **kwargs):
+        super(AnonUser, self).save(*args, **kwargs)
```
Comment (eivl) [inline whimsical-woodpeckers/anon/models.py:33]: Nice to see the usage of *args and **kwargs with super()

File: whimsical-woodpeckers/anon/views.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+# Create your views here.
+
+
+def random_string(length):
+    i = 0
+    pw = ""
+    while i < length:
+        pw = pw + random.choice(string.printable)
+        i += 1
+    return pw
```
Comment (eivl) [inline whimsical-woodpeckers/anon/views.py:17]: random.choice does not scale well. it would be better to use random.choices here.

File: whimsical-woodpeckers/messages/consumers.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        self.user = None
+        self.groups = set()
+
+    def disconnect(self, close_code):
+        UserChannels.objects.filter(name=self.channel_name).delete()
+
+    def group_name(self, user):
+        return f"messaging-{self.user.id}"
+
+    def receive(self, text_data):
```
Comment (eivl) [inline whimsical-woodpeckers/messages/consumers.py:26]: Feels like you are doing a lot of things inside the function `receive` It might have been wise to refactor some business logic out.

File: whimsical-woodpeckers/messages/views.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# from django.shortcuts import render
```
Comment (eivl) [inline whimsical-woodpeckers/messages/views.py:1]: should be no need to commit a file with all lines commented out.

File: whimsical-woodpeckers/www/src/py/common/__init__.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# I really, really wanted to do from enum import Enum
+# Alas, transcrypt won't import from the standard library
+
+
+class AwfulEnum:    # I'm so sorry
```
Comment (eivl) [inline whimsical-woodpeckers/www/src/py/common/__init__.py:5]: Im not sure what you mean, but I do feel you're very sorry for this.

File: whimsical-woodpeckers/www/src/py/components/directory/__init__.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        def auth_callback(data):
+            parent_this.token = data['token']
+            callback()
+
+        Fetch.get("data/") \
+            .then(lambda response: response.json()) \
+            .then(lambda data: data_callback(data))
+        Fetch.get("auth/") \
+            .then(lambda response: response.json()) \
+            .then(lambda data: auth_callback(data))
```
Comment (eivl) [inline whimsical-woodpeckers/www/src/py/components/directory/__init__.py:62]: This looks very overengineered. Method chaining using lambdas. This is either brilliant or terrible. I'm not really sure what. ill edge on it being brilliant.

File: whimsical-woodpeckers/www/src/py/fetch.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def dict_to_iterdict(thing):
+    pls = {}
+    for i in thing:
+        pls[i] = thing[i]
+    return pls
+
+
+class Fetch:
```
Comment (eivl) [inline whimsical-woodpeckers/www/src/py/fetch.py:19]: some docstrings would have been nice here to share whats going on.

File: whimsical-woodpeckers/www/src/py/pyvue.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
+# __pragma__ ('js', "{}", "import * as VuePkg from 'vue';")
```
Comment (eivl) [inline whimsical-woodpeckers/www/src/py/pyvue.py:1]: the whole pyvue context should have been documented with docstrings. It's very hard to follow along. 
The use of advanced python concepts is nice to see though.

File: whimsical-woodpeckers/www/src/py/websocket.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
+# __pragma__ ('skip')
+window = JSON = this = js_websocket = __new__ = 0    # Prevent complaints by optional static checker
+# __pragma__ ('noskip')
+
+
+# __pragma__("kwargs")
+class WebSocket:
```
Comment (eivl) [inline whimsical-woodpeckers/www/src/py/websocket.py:7]: nice and clear, but why all the `pass` in the methods?

File: whimsical-woodpeckers/www/src/scss/XP.css-main/README.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
+## XP.css
```
Comment (eivl) [inline whimsical-woodpeckers/www/src/scss/XP.css-main/README.md:1]: nice that you added the licensing information for this third party theme.

**Review by eivl — COMMENTED** [review]
The project uses node and pipenv. The readme documents the commands needed to be run to setup the project. except the missing migrate command, it did work on my end. But alas I was not able to get a working version on screen. 

Maybe third party assets should have been mentioned in the README as well. 
This project looks very advanced. with python and vue. details as using `*args` and `**kwargs` alongside super() is nice to see. The overly advance coding is in fact nice to see, but having some docstrings along the way would make it easier to understand. Some functions are a bit fat, and might justify some refactoring or a well written docstring explaining the need. 

The Vue part looks very advanced, and beside a few comments its missing some much wanted docstring that explains whats going on. Without this kind of explanation the advanceness of the code just looks a bit overengineered. 

Im a bit sad I could not get a working version of this on my system, but the whole message sending over sockets looks really well written. I would love to test this if the team wouldn't mind setting up a ngrok for me. 

I wish this team all the best as a possibility of a high ranking is definivly possible.
